staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -113,71 +114,132 @@ func WithPriorityAndFairness(
-							const size = 64 << 10
-							buf := make([]byte, size)
-							buf = buf[:runtime.Stack(buf, false)]
-							err = fmt.Sprintf("%v\n%s", err, buf)
-						}
-						resultCh <- err
-					}()
-
-					// Protect from the situations when request will not reach storage layer
-					// and the initialization signal will not be send.
-					defer watchInitializationSignal.Signal()
-
-					handler.ServeHTTP(w, innerReq)
-				}()
+				forgetWatch = fcIfc.RegisterWatch(requestInfo)
+
+				// Notify the main thread that we're ready to start the watch.
+				close(shouldStartWatchCh)
 
+				// Wait until the request is finished from the APF point of view
+				// (which is when its initialization is done).
 				watchInitializationSignal.Wait()
-			} else {
-				handler.ServeHTTP(w, innerReq)
 			}
-		}
 
-		// find the estimated "width" of the request
-		// TODO: Maybe just make it costEstimator and let it return additionalLatency too for the watch?
-		// TODO: Estimate cost should also take fcIfc.GetWatchCount(requestInfo) as a parameter.
-		width := widthEstimator.EstimateWidth(r)
-		digest := utilflowcontrol.RequestDigest{RequestInfo: requestInfo, User: user, Width: width}
+			// Ensure that an item can be put to resultCh asynchronously.
+			resultCh := make(chan interface{}, 1)
 
-		fcIfc.Handle(ctx, digest, note, func(inQueue bool) {
-			if inQueue {
-				noteWaitingDelta(1)
-			} else {
-				noteWaitingDelta(-1)
+			// Call Handle in a separate goroutine.
+			// The reason for it is that from APF point of view, the request processing
+			// finishes as soon as watch is initialized (which is generally orders of
+			// magnitude faster then the watch request itself). This means that Handle()
+			// call finishes much faster and for performance reasons we want to reduce
+			// the number of running goroutines - so we run the shorter thing in a
+			// dedicated goroutine and the actual watch handler in the main one.
+			go func() {
+				defer func() {
+					err := recover()
+					// do not wrap the sentinel ErrAbortHandler panic value
+					if err != nil && err != http.ErrAbortHandler {
+						// Same as stdlib http server code. Manually allocate stack
+						// trace buffer size to prevent excessively large logs
+						const size = 64 << 10
+						buf := make([]byte, size)
+						buf = buf[:runtime.Stack(buf, false)]
+						err = fmt.Sprintf("%v\n%s", err, buf)
+					}
+
+					// Ensure that the result is put into resultCh independently of the panic.
+					resultCh <- err
+				}()
+
+				// We create handleCtx with explicit cancelation function.
+				// The reason for it is that Handle() underneath may start additional goroutine
+				// that is blocked on context cancellation. However, from APF point of view,
+				// we don't want to wait until the whole watch request is processed (which is
+				// when it context is actually cancelled) - we want to unblock the goroutine as
+				// soon as the request is processed from the APF point of view.
```

Comment:
APF doesn't care how many parallel watchers you establish? I would have expected that to matter to APF, and telling APF the request has finished before it actually has makes it impossible for APF to track that, right?

Reply:
It's hard to explain in few words. if you're interested in more details it's in these two keps:
https://github.com/kubernetes/enhancements/pull/2669/files
https://github.com/kubernetes/enhancements/pull/2700

But trying to put it short, we divide watch into two phases:
1) initialization
2) keeping up with incoming events

(1) is what we handle here exactly [and we finish the goroutine as soon as watch is initialized]
(2) after that, the cost of watch is kind-of connected to the write request, for multiple different reasons described in the KEP

Reply:
> after that, the cost of watch is kind-of connected to the write request, for multiple different reasons described in the KEP

but who is the cost attributed to, the watcher(s) or the writer? I would expect APF to prevent a user from opening a ton of parallel watches rather than keep a writer from doing a single write to a resource that happens to have a lot of watches on it.

Reply:
we can hoist this out of this PR... it's clearly a bigger discussion... it just seems like this PR is wiring up stuff that will have to be undone to get APF to behave the way I (probably naively) expected it to

Reply:
> it just seems like this PR is wiring up stuff that will have to be undone to get APF to behave the way I (probably naively) expected it to

No - the initialization will remain as is [the initialization is kind-of done as no-longrunning requests - from POV of APF the watch request finishes after initialization is done.]

Re "keeping watch up-to-date" (which is a separate problem) - let's take it out of this PR [it's an orthogonal discussion.]